concent_api/core/payments/storage.py

```diff
@@ -0,0 +1,150 @@
+import logging
+from typing import Dict
+from typing import List
+from hexbytes import HexBytes
+from ethereum.transactions import Transaction
+from golem_sci.transactionsstorage import TransactionsStorage
+
+from django.db import transaction
+
+from core.models import GlobalTransactionState
+from core.models import PendingEthereumTransaction
+
+from core.exceptions import TransactionNonceNotMatchError
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseTransactionsStorage(TransactionsStorage):
+    """
+    Concent custom implementation of TransactionsStorage interface used to store Ethereum transaction data into
+    database using Django models.
+    """
+
+    def __init__(self, nonce: int, *args: List, **kwargs: Dict) -> None:
+        super().__init__(*args, **kwargs)
+
+        assert isinstance(nonce, int)
+        if not GlobalTransactionState.objects.filter(pk=0).exists():
+            global_transaction_state = GlobalTransactionState(
+                pk=0,
+                nonce=nonce,
+            )
+            global_transaction_state.full_clean()
+            global_transaction_state.save()
+
+    @transaction.atomic(using='control')
+    def get_nonce(self) -> int:
+        """
+        Returns the nonce for the next transaction.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+        return global_transaction_state.get_nonce
+
+    @transaction.atomic(using='control')  # pylint: disable=no-self-use
+    def get_all_tx(self) -> List[Transaction]:
+        """
+        Returns the list of all transactions.
+        """
+        return [
+            Transaction(
+                nonce=ethereum_transaction.get_nonce,
+                gasprice=ethereum_transaction.get_gasprice,
+                startgas=ethereum_transaction.get_startgas,
+                value=ethereum_transaction.get_value,
+                v=ethereum_transaction.v,
+                r=ethereum_transaction.get_r,
+                s=ethereum_transaction.get_s,
+                data=ethereum_transaction.data.tobytes(),
+                to=HexBytes(ethereum_transaction.to.tobytes()).hex(),
+            )
+            for ethereum_transaction in PendingEthereumTransaction.objects.all()
+        ]
+
+    @transaction.atomic(using='control')
+    def put_tx_and_inc_nonce(self, tx: Transaction) -> None:
+        """
+        Save a valid transaction and increase the nonce.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+
+        if global_transaction_state.get_nonce != tx.nonce:
+            raise TransactionNonceNotMatchError(
```

Comment:
Maybe `TransactionNonceMismatch` would be a better name?

Reply:
Done